frontend/src/metabase-lib/queries/utils/dataset.ts

```diff
@@ -53,21 +54,32 @@ export function findColumnIndexForColumnSetting(
 }
 
 export function findColumnSettingIndexForColumn(
+  query: Lib.Query,
   columnSettings: TableColumnOrderSetting[],
   column: DatasetColumn,
 ) {
-  const fieldRef = fieldRefForColumn(column);
-  const normalizedFieldRef = fieldRef ? normalizeFieldRef(fieldRef) : undefined;
-  if (normalizedFieldRef == null) {
-    return columnSettings.findIndex(
-      columnSetting => columnSetting.name === column.name,
-    );
-  }
-  const index = columnSettings.findIndex(
-    columnSetting =>
-      columnSetting.fieldRef &&
-      _.isEqual(normalizedFieldRef, normalizeFieldRef(columnSetting.fieldRef)),
+  // ignore settings without fieldRef but preserve indexes
+  const items = columnSettings.reduce<
+    { fieldRef: FieldReference; settingIndex: number }[]
+  >((items, { fieldRef }, settingIndex) => {
+    if (fieldRef != null) {
+      items.push({ fieldRef, settingIndex });
+    }
+    return items;
+  }, []);
```

Comment:
This is a bit cleaner, does not need explicit typing and helps avoid a cast later on when calling `Lib.findColumnIndexesFromLegacyRefs` (we can just use `items.map(({ fieldRef }) => fieldRef)`):

```suggestion
  const items = columnSettings.flatMap((item, settingIndex) => {
    const fieldRef = item.fieldRef ? normalizeFieldRef(item.fieldRef) : null;
    return fieldRef ? [{ fieldRef, settingIndex }] : [];
  });
```

Reply:
[6f121a1](https://github.com/metabase/metabase/pull/37856/commits/6f121a1bd4dab41efac91abca2c06b082578b505)